### ISSUE TITLE
False CS8602 is raised by Roslyn when an IEnumerable<T> is implied to be null

### ISSUE THREAD

**Issue by bornachk** [dotnet/roslyn#78635]
**Version Used**: .NET SDK 9.0.203, C# language version 13.0

**Steps to Reproduce**:

1. Create a .NET class library project (or compile/analyze this code with your own means, while showing warnings)
2. Create a non-nullable `IEnumerable<T>` object of any `T` type or size (a derivative like `T[]` or `List<T>` can also be used)
3. Reference that object using a redundant null conditional operator (i.e. `obj?.ToString()`)
4. Reference that object without using a null conditional operator (i.e. `obj.ToString()`)

**Diagnostic Id**: `CS8602: Dereference of a possibly null reference.`

**Expected Behavior**: CS8602 should not be raised in this scenario.

**Actual Behavior**: CS8602 is raised in this scenario.

**Example Code**:
```csharp
namespace shutupcompiler;

public class Class1
{
    public static void DoSomething()
    {
        // With a non-nullable array, implying that it is nullable
        // will cause the code after to result in null reference warnings:
        object[] nonNullableArr = []; // Create an array with zero elements
                                      // (the way it is created is irrelevant)

        // These alternative construction methods still not change the analysis/output either:
        // object[] nonNullableArr = new object[0];
        // Or the following grotesque method will still not change the analysis
        // object[] nonNullableArr = [.. new object().GetType().Assembly.CustomAttributes.Where(_ => false).Cast<object>()];
        nonNullableArr?.ToString(); // This implies arr is a nullable type, though it is not.
        nonNullableArr.ToString();  // CS8602: Dereference of a possibly null reference.
                                    // Despite arr's target type not being nullable,
                                    // this erroneous warning will appear.

        // With a blank slate:
        object[] anotherArr = nonNullableArr;
        anotherArr.ToString();      // An object reference to the same array with
                                    // the same action does not cause the warning.

        // This error only manifests with strictly IEnumerable<T>.
    }
}
```

Many other similar issues pertaining to this issue might come from the same underlying issue, but those show a bug that occurs on a **nullable** object of any type (or a specific type). This bug occurs specifically in `IEnumerable<T>` and derived types.

**Comment by RikkiGibson** [comment]
Thanks for filing this issue. The behavior you are observing appears to be by design.

- Any variable of reference type can get a maybe-null flow state through null tests. It is thought to be desirable that `if (nonNullableArr is null) nonNullableArray.ToString();` results in a warning, for example. The annotation (presence or absence of `?`) mostly just affects the variable's initial state in flow analysis, and whether warnings are reported for *assigning* possible null values to it.
- As you saw, `a?.b` can cause `a` to transition from not-null to maybe-null flow state. It is thought to be desirable that if user code indicates a suspicion that the value is null, that we begin treating the value as maybe-null at that point.
- At the same time--a warning on a dereference `a.b` actually causes the value `a` to transition from maybe-null to not-null. This is mainly because the code following the dereference is *only reachable when the value was not-null*--otherwise an exception would be thrown.

**Comment by bornachk** [comment]
Now that I see these points, I get why this behavior is like this. Still, there are some inconsistencies:
> if there is a null check the variable can apparently be null.
But passing the variable to another variable with an object reference will reset the possibility of it being null. Not exactly sure how the internals of Roslyn is implemented but this seems possible.
Take an implicitly null array using this method 
```
csharp
object[] arr = [];
arr?.ToString(); // marks as nullable?
object[] copy = arr;
arr.ToString(); // will not generate a warning despite it inherently being a reference to a potentially null object 
```

**Comment by RikkiGibson** [comment]
I am seeing warnings in that scenario: [.NET Lab](https://lab.razor.fyi/#fVDBSkJBFCVyUbOKvuAuddFTWkQoIqEQRUGgq0ToOl5fk_Nm7N55xiP8n_6gvwj6gJZ9Q9t4vhQz7O7uOeeec2bUa0mpG_YxYxJpOXwppWJcDN1MAiUNtb5FbW8t6WC8k-icHLHR_ygukiQNOLS0oekYjJ2XYLRsZ6K2H9GZQ5uJ2ZRdGfe4AfXumXBkXLwNj3ooE2kopS2KwM-L1bMCAJCAwWiYeTOCazSuXFnABZmPHz6QDv0BIDM0oT9orChkbkU93w1sXFyuNKBahQR5IoACLrU2_4HWXyftpxk08_PFyROyy3u3u6cntVod2t7NiEMO5SZgTSBGC55h6kXM0FJBzNCmBMGD8-5omQchm1K03nGz4q-84zp0iGlMTE4T-DHgMiUrUlZkYTpXc3W5y6m73d97e__6_KCD0t3ONw). There is a warning on each of the last 2 lines.
